format/bplist/bplist.go

```diff
@@ -0,0 +1,246 @@
+	elementTypeReal:             {Sym: "real", Description: "Floating Point Number"},
+	elementTypeDate:             {Sym: "date", Description: "Date, 4 or 8 byte float"},
+	elementTypeData:             {Sym: "data", Description: "Binary data"},
+	elementTypeASCIIString:      {Sym: "ascii_string", Description: "ASCII encoded string"},
+	elementTypeUnicodeString:    {Sym: "unicode_string", Description: "Unicode string"},
+	elementTypeUID:              {Sym: "uid", Description: "UID"},
+	elementTypeArray:            {Sym: "array", Description: "Array"},
+	elementTypeSet:              {Sym: "set", Description: "Set"},
+	elementTypeDict:             {Sym: "dict", Description: "Dictionary"},
+}
+
+// decodes the number of bits required to store the following object
+func decodeSize(d *decode.D, sms ...scalar.Mapper) uint64 {
+	n := d.FieldU4("size_bits")
+	if n != 0x0f {
+		return n
+	}
+
+	d.FieldU4("large_size_marker", d.AssertU(0b0001))
+
+	// get the exponent value
+	n = d.FieldU4("exponent")
+
+	// calculate the number of bytes encoding the size
+	n = 1 << n
+
+	// decode that many bytes as big endian
+	n = d.FieldUFn(
+		"size_bigint",
+		func(d *decode.D) uint64 {
+			v := d.UBigInt(int(n * 8))
+			d.AssertBigIntRange(big.NewInt(1), big.NewInt(math.MaxInt64))
+			return v.Uint64()
+		}, sms...)
+
+	return n
+}
+
+func decodeItem(d *decode.D, p *plist) {
+	m := d.FieldU4("type", elementTypeMap)
+	switch m {
+	case elementTypeNullOrBoolOrFill:
+		t := d.U4()
+		switch t {
+		case null:
+			d.FieldValueNil("value")
+		case boolTrue:
+			d.FieldValueBool("value", true)
+		case boolFalse:
+			d.FieldValueBool("value", false)
+		}
+	case elementTypeInt:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		if n*8 <= 64 {
+			d.FieldU("value", int(n*8))
+		} else {
+			d.FieldUBigInt("value", int(n))
+		}
+	case elementTypeReal:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldF("value", int(n))
+	case elementTypeDate:
+		n := 1 << decodeSize(d, d.AssertU(4, 8))
+		d.FieldValueU("size", uint64(n))
+		d.FieldF("value", n*8, scalar.DescriptionActualFCocoaDate)
+	case elementTypeData:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldRawLen("value", int64(n*8))
+	case elementTypeASCIIString:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldUTF8("value", int(n))
+	case elementTypeUnicodeString:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldUTF16("value", int(n))
+	case elementTypeUID:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldUBigInt("value", int(n)).Uint64()
+	case elementTypeArray:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		i := uint64(0)
+		d.FieldStructArrayLoop("entries", "entry",
+			func() bool { return i < n },
+			func(d *decode.D) {
+				idx := d.FieldU8("object_index")
+				decodeReference(d, p, idx)
+				i++
+			})
+	case elementTypeSet:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		i := uint64(0)
+		d.FieldStructArrayLoop("entries", "entry",
```

Comment:
In some other decoders i've skipped using the `d.FieldStrctArrayLoops()` and instead use `for` as I think it feels a bit awkward. Any idea how to make it nicer? maybe a specialized version for given number of entries? maybe with generics it can be made nicer?

Reply:
But i'm fine with this also, it's easy to find and possibly refactor in the future

Reply:
Added a function `FieldStructNArray` for decoding arrays with a known number of elements, does that look okay?